File: Smart Sweepers/src/utils/Timer.java
package utils;

public class Timer {

	long currentTime;
	long lastTime;
	long nextTime;
	long frameTime;

	double timeElapsed;

	float fps;

	public Timer(float fps) {
		if (fps == 0) {
			throw new IllegalArgumentException("FPS of 0 is not allowed");
		}
		timeElapsed = 0;
		lastTime = 0;
		this.fps = fps;

		frameTime = (long) ((1 / fps) * 1_000_000_000);
	}

	public void start() {
		lastTime = System.nanoTime();

		nextTime = lastTime + frameTime;
	}

	public boolean readyForNextFrame() {
		currentTime = System.nanoTime();

		if (currentTime > nextTime) {
			timeElapsed = (currentTime - lastTime);
			lastTime = currentTime;

			nextTime = currentTime + frameTime;
			return true;
		}
		return false;
	}

	public double timeElapsed() {
		currentTime = System.nanoTime();

		timeElapsed = (currentTime - lastTime);
		lastTime = currentTime;

		return timeElapsed;
	}

	public double getTimeElapsed() {
		return timeElapsed;
	}
}

File: Smart Sweepers/src/utils/Timer.kt
package utils

class Timer(fps: Float) {
    init {
        require(!(fps <= 0.0)) { "FPS of 0 is not allowed" }
    }

    private var currentTime = 0L
    private var lastTime = 0L
    private var nextTime = 0L
    private val frameTime = ((1 / fps) * 1_000_000_000).toLong()
    private var timeElapsed = 0.0

    private fun resetLastAndNextTimesFrom(newLast: Long) {
        lastTime = newLast
        nextTime = newLast + frameTime
    }

    fun start() = resetLastAndNextTimesFrom(System.nanoTime())

    fun readyForNextFrame(): Boolean {
        val now = System.nanoTime().also { currentTime = it }
        return (now > nextTime).also { passedNext ->
            if (passedNext) {
                timeElapsed = (now - lastTime).toDouble()
                resetLastAndNextTimesFrom(now)
            }
        }
    }
}

